### PR TITLE
Added missing internal component conversion for component types

### PR DESCRIPTION
**Pull request by JCash**
Forum post:
https://forum.defold.com/t/defold-1-9-0-beta/77189/15?u=mathias_westerdahl

## Technical details
The issue was that refactoring the use of dmGameObject::GetComponentFromLua requires proper conversion from dmGameObject::HComponentInternal to the actual pointer. Some component types lacked this conversion function.
I've added a test that checks for this, should the components' internal representation change.

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/gamesys/src/gamesys/components/comp_camera.cpp [hunk 1]
```diff
@@ -179,6 +179,11 @@ namespace dmGameSystem
         return dmGameObject::CREATE_RESULT_OK;
     }
 
+    void* CompCameraGetComponent(const dmGameObject::ComponentGetParams& params)
+    {
+        return (void*)params.m_UserData;
+    }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_camera.cpp:185]: The camera already returned the pointer from the `CompCameraCreate()`function, so no conversion needed.
While I didn't need to add this function to the camera, I though it better to add it to the most common component types. We can later add for the rest of them as well, but I ran out of time. Also, the tests checks some that don't have the function (e.g. collectionproxy)

File: engine/gamesys/src/gamesys/components/comp_factory.h [hunk 2]
```diff
@@ -20,14 +20,15 @@
...
+    dmGameObject::CreateResult      CompFactoryNewWorld(const dmGameObject::ComponentNewWorldParams& params);
+    dmGameObject::CreateResult      CompFactoryDeleteWorld(const dmGameObject::ComponentDeleteWorldParams& params);
+    dmGameObject::CreateResult      CompFactoryCreate(const dmGameObject::ComponentCreateParams& params);
+    dmGameObject::CreateResult      CompFactoryDestroy(const dmGameObject::ComponentDestroyParams& params);
+    dmGameObject::CreateResult      CompFactoryAddToUpdate(const dmGameObject::ComponentAddToUpdateParams& params);
+    dmGameObject::UpdateResult      CompFactoryUpdate(const dmGameObject::ComponentsUpdateParams& params, dmGameObject::ComponentsUpdateResult& update_result);
+    dmGameObject::UpdateResult      CompFactoryOnMessage(const dmGameObject::ComponentOnMessageParams& params);
+    dmGameObject::PropertyResult    CompFactoryGetProperty(const dmGameObject::ComponentGetPropertyParams& params, dmGameObject::PropertyDesc& out_value);
+    void*                           CompFactoryGetComponent(const dmGameObject::ComponentGetParams& params);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_factory.h:31]: Indentation, except for this last function.

File: engine/gamesys/src/gamesys/components/comp_mesh.cpp [hunk 3]
```diff
@@ -396,6 +396,12 @@ namespace dmGameSystem
         return dmGameObject::CREATE_RESULT_OK;
     }
 
+    void* CompMeshGetComponent(const dmGameObject::ComponentGetParams& params)
+    {
+        MeshWorld* world = (MeshWorld*)params.m_World;
+        return (void*)world->m_Components.Get(params.m_UserData);
+    }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_mesh.cpp:403]: Some components missed their "GetComponent" function, and need to convert from a `dmGameObject::HComponentInternal` handle to an actual pointer.

File: engine/gamesys/src/gamesys/components/comp_model.cpp [hunk 4]
```diff
@@ -727,6 +727,12 @@ namespace dmGameSystem
         return dmGameObject::CREATE_RESULT_OK;
     }
 
+    void* CompModelGetComponent(const dmGameObject::ComponentGetParams& params)
+    {
+        ModelWorld* world = (ModelWorld*)params.m_World;
+        return (void*)world->m_Components.Get(params.m_UserData);
+    }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_model.cpp:734]: The reported issue was related to this missing function.

File: engine/gamesys/src/gamesys/test/script/test_component.script [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+-- https://www.defold.com/license
+-- 
+-- Unless required by applicable law or agreed to in writing, software distributed
+-- under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+-- CONDITIONS OF ANY KIND, either express or implied. See the License for the
+-- specific language governing permissions and limitations under the License.
+
+function init(self)
+    test_callback(test_url)
+end
```
Comment (JCash) [inline engine/gamesys/src/gamesys/test/script/test_component.script:17]: We call back into C++, which holds the data, and we can call `dmGameObject::GetComponentFromLua()` and check that it's not an index.

File: engine/gamesys/src/gamesys/test/test_gamesys.cpp [hunk 6]
```diff
@@ -834,6 +835,79 @@ TEST_F(BufferMetadataTest, MetadataLuaApi)
...
+{
+    lua_getglobal(L, "test_data");
+    ScriptComponentTestData* data = (ScriptComponentTestData*)lua_touserdata(L, -1);
+    lua_pop(L, 1);
+
+    dmGameObject::HComponent out_component = 0;
+    dmGameObject::GetComponentFromLua(L, 1, data->m_Collection, data->m_ComponentType, &out_component, 0, 0);
+
+    // We should have an actual pointer at this stage, and it is not likely it is less than a certain low number
+    lua_pushboolean(L, (uintptr_t)out_component > 100000);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/test/test_gamesys.cpp:854]: The test is simple.
We don't check for validity per se, but rather if it looks to be an index or not.

**Review by AGulev — APPROVED** [review]
👍

**Pull request merged**